core/drivers/clk/clk-stm32mp25.c

```diff
@@ -0,0 +1,3449 @@
+		 GATE_APB2DIV_RDY),
+	_DIV_CFG(DIV_APB3, RCC_APB3DIVR, 0, 3, 0, apb_div_table,
+		 GATE_APB3DIV_RDY),
+	_DIV_CFG(DIV_APB4, RCC_APB4DIVR, 0, 3, 0, apb_div_table,
+		 GATE_APB4DIV_RDY),
+	_DIV_CFG(DIV_APBDBG, RCC_APBDBGDIVR, 0, 3, 0, apb_div_table,
+		 GATE_APBDBGDIV_RDY),
+	_DIV_CFG(DIV_LSMCU, RCC_LSMCUDIVR, 0, 1, 0, NULL, GATE_LSMCUDIV_RDY),
+};
+
+enum stm32_osc {
+	OSC_HSI,
+	OSC_HSE,
+	OSC_MSI,
+	OSC_LSI,
+	OSC_LSE,
+	NB_OSCILLATOR
+};
+
+struct clk_stm32_bypass {
+	uint16_t offset;
+	uint8_t bit_byp;
+	uint8_t bit_digbyp;
+};
+
+struct clk_stm32_css {
+	uint16_t offset;
+	uint8_t bit_css;
+};
+
+struct clk_stm32_drive {
+	uint16_t offset;
+	uint8_t drv_shift;
+	uint8_t drv_width;
+	uint8_t drv_default;
+};
+
+struct clk_oscillator_data {
+	const char *name;
+	unsigned long frequency;
+	uint16_t gate_id;
+	struct clk_stm32_bypass *bypass;
+	struct clk_stm32_css *css;
+	struct clk_stm32_drive *drive;
+};
+
+#define BYPASS(_offset, _bit_byp, _bit_digbyp) \
+	(&(struct clk_stm32_bypass){\
+		.offset = (_offset),\
+		.bit_byp = (_bit_byp),\
+		.bit_digbyp = (_bit_digbyp),\
+	})
+
+#define CSS(_offset, _bit_css) \
+	(&(struct clk_stm32_css){\
+		.offset = (_offset),\
+		.bit_css = (_bit_css),\
+	})
+
+#define DRIVE(_offset, _shift, _width, _default) \
+	(&(struct clk_stm32_drive){\
+		.offset = (_offset),\
+		.drv_shift = (_shift),\
+		.drv_width = (_width),\
+		.drv_default = (_default),\
+	})
+
+#define OSCILLATOR(idx_osc, _name, _gate_id, _bypass, _css, _drive) \
+	[(idx_osc)] = (struct clk_oscillator_data){\
+		.name = (_name),\
+		.gate_id = (_gate_id),\
+		.bypass = (_bypass),\
+		.css = (_css),\
+		.drive = (_drive),\
+	}
+
+static struct clk_oscillator_data stm32mp25_osc_data[NB_OSCILLATOR] = {
+	OSCILLATOR(OSC_HSI, "clk-hsi", GATE_HSI,
+		   NULL, NULL, NULL),
+
+	OSCILLATOR(OSC_LSI, "clk-lsi", GATE_LSI,
+		   NULL, NULL, NULL),
+
+	OSCILLATOR(OSC_MSI, "clk-msi", GATE_MSI,
+		   NULL, NULL, NULL),
+
+	OSCILLATOR(OSC_LSE, "clk-lse", GATE_LSE,
+		   BYPASS(RCC_BDCR, RCC_BDCR_LSEBYP_BIT,
+			  RCC_BDCR_LSEDIGBYP_BIT),
+		   CSS(RCC_BDCR, RCC_BDCR_LSECSSON_BIT),
+		   DRIVE(RCC_BDCR, RCC_BDCR_LSEDRV_SHIFT,
+			 RCC_BDCR_LSEDRV_WIDTH, LSEDRV_MEDIUM_HIGH)),
+
+	OSCILLATOR(OSC_HSE, "clk-hse", GATE_HSE,
+		   BYPASS(RCC_OCENSETR, RCC_OCENSETR_HSEBYP_BIT,
+			  RCC_OCENSETR_HSEDIGBYP_BIT),
+		   CSS(RCC_OCENSETR, RCC_OCENSETR_HSECSSON_BIT),
+		   NULL),
+};
+
+static struct clk_oscillator_data *clk_oscillator_get_data(unsigned int osc_id)
+{
+	assert(osc_id < ARRAY_SIZE(stm32mp25_osc_data));
+
+	return &stm32mp25_osc_data[osc_id];
+}
+
+static unsigned long clk_stm32_get_rate_oscillator(unsigned int osc_id)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[osc_id];
+
+	return osci->freq;
+}
+
+static unsigned long clk_stm32_pll_get_oscillator_rate(unsigned int sel)
+{
+	int osc[] = { OSC_HSI, OSC_HSE, OSC_MSI };
```

Comment:
s/int/unsigned int/ since it's the argument type `clk_stm32_get_rate_oscillator()` expects.